### PR TITLE
Keycloak oidc provider

### PR DESCRIPTION
**Pull request by pb82**
## Description

Keycloak Provider based on OIDC functionality

supersedes #1107 

## Motivation and Context

Keycloak is OIDC and we should deprecate the OAuth based keycloak provider that supports limited functionality.

## How Has This Been Tested?

Unit Tests - Keycloak community testing & feedback desired.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/keycloak_oidc_test.go
Comment (NickMeves) [inline providers/keycloak_oidc_test.go:42]: For any new test files, we are looking to migrate everything to using Ginkgo BDD test suites.

As a reference, you can keep most of these table driven tests, here's a sample from @JoelSpeed using the `DescribeTable` functionality in Ginkgo: https://github.com/oauth2-proxy/oauth2-proxy/pull/1142/files

If you need any references, just look at any of @JoelSpeed 's PRs, he's a Ginkgo testing wizard 😄

**Comment by NickMeves** [comment]
@JoelSpeed since @pb82 is taking this PR (https://github.com/oauth2-proxy/oauth2-proxy/pull/1107) I started to the finish line with a test suite and I'm assuming live validation on an actual Keycloak setup:

Me reviewing would be a little incestuous 😅 - reviewing my own implementation code. Do you mind taking on reviewing this?

File: providers/keycloak_oidc.go
Comment (JoelSpeed) [inline providers/keycloak_oidc.go:115]: I probably wouldn't log here, I'd be concerned that this could end up being logged very frequently. Until we have the ability to distinguish multiple logging levels, I'd prefer not to add this kind of debug log

**Review by JoelSpeed — COMMENTED** [review]
Nothing too much to add here, though @NickMeves if you could review my comments that would be great

Thanks for your work so far @pb82

File: pkg/apis/options/legacy_options.go
Comment (NickMeves) [reply pkg/apis/options/legacy_options.go:511]: > Perhaps we want to keep it generic so that is can be used by multiple providers in the future 🤔

That was my intent - for commonly named things, keep the naming generic so it is easy to configure & document if multiple providers could potentially implement (this moves away from some of the conventions past providers have used, though).

I think this makes the transition to structured configs when this setting would be a child of a provider settings block a little clearer -- can keep the same name and not need to remove a provider specific prefix.

File: providers/keycloak_oidc.go
Comment (NickMeves) [inline providers/keycloak_oidc.go:62]: @JoelSpeed Something I changed here that deviates from a convention we started in the `gitlab` provider that I think I want to change as well:

When there are multiple group/role types to populate our session.Groups list -- only things that aren't explicitly already a `group` will get a prefix (e.g. `role:blah`). Leaving the groups raw instead of `group:somegroup`.

This way, we can fully rely the base `OIDCProvider` logic for standard groups extraction without having to make each Provider we use this new pattern of extending it have to override that base logic.

Comment (NickMeves) [reply providers/keycloak_oidc.go:62]: If you concur, before we have the next major release, I'd like to sneak into the `gitlab` provider and make this shift.  I think the `gitlab` provider could use a refactor to be based on the `OIDCProvider` like this PR to remove alot of duplicate code it has.

File: providers/keycloak_oidc_test.go
Comment (pb82) [reply providers/keycloak_oidc_test.go:42]: Thanks, yes I should rewrite them using the newer style. Looks like some examples are already in the repo.

File: providers/keycloak_oidc.go
Comment (pb82) [reply providers/keycloak_oidc.go:47]: :+1:

File: providers/keycloak_oidc.go
Comment (pb82) [reply providers/keycloak_oidc.go:115]: :+1:

File: pkg/apis/options/legacy_options.go
Comment (pb82) [reply pkg/apis/options/legacy_options.go:511]: I agree, `allowed-roles` does not have to be keycloak specific, other providers could also make use of it. Is that ok for you @JoelSpeed or would you prefer it to be prefixed with keycloak?

**Comment by pb82** [comment]
@JoelSpeed @NickMeves I have it all working in a local environment. I'll migrate the tests to the new style. Do you want me to add docs as well?

ping @tpohl

**Comment by pb82** [comment]
Tests migrated to Gingko, Docs added and Changelog updated.

**Comment by pb82** [comment]
@JoelSpeed can you please run another build?

File: pkg/apis/options/legacy_options.go
Comment (JoelSpeed) [reply pkg/apis/options/legacy_options.go:511]: My only concern here is that we should document which providers it works for, otherwise we will be in a situation as with many of our issues, where people try the feature only to realise it doesn't work with their provider. Let's make an issue to make sure the provider features are tallied up and documented per provider

File: providers/keycloak_oidc.go
Comment (JoelSpeed) [reply providers/keycloak_oidc.go:61]: Are there any providers that we know of that allow `:` as a character in the group name, if there are, then a user could theoretically create a group called `role:admin`, which would then clash with a role called `admin`. If we explicitly prefix within  the groups as well, I don't think this confusion is possible.

I'm concerned this will one day turn into a weird bug where a user ends up in a role because of an ill named group :thinking:

**Comment by JoelSpeed** [comment]
@pb82 Apologies as this isn't really anything you can directly fix, but I'd like to have a discussion around https://github.com/oauth2-proxy/oauth2-proxy/pull/1210#discussion_r644261689 before we merge. I have some concerns about this pattern so will probably need @NickMeves to weigh in here as he did the original draft of this

File: pkg/apis/options/legacy_options.go
Comment (tpohl) [reply pkg/apis/options/legacy_options.go:511]: Hi, I understand your concerns and it might pose a risk of unintended behavior.
Still I think that especially for Keycloak there is a difference between group membership and roles (realm and client level). 

In our usecase we use roles for permissions -> for this proxy this means „you can pass or you shall not pass“ and we use groups as a manner of managing users (all members of a group have the same permissions but multiple groups can have those permissions). 

I have seen other users and implementations that use the „groups“ for permissions. 

On top of that at least Keycloak will allow us to design the tokens the way we want to… so there is a lot that can be done wrong.

The implementation of @pb82 leverages the default behavior of Keycloak and where it would map groups and roles into the token. So I think for Keycloak it is alright to use „role:“ as a delimiter. But: Keycloak does allow me to create a group „role:admin“ and I am not aware of any way around this. 
One Idea would be to configure if we want to authorize based upon groups or on roles.

One extr thing, that I could imagine is that, if we went for a completely independent OIDC JWT implementation we could configure a JSON Path to the claim in the token, we want to verify and then configure what values would be required to let the user pass.

File: providers/keycloak_oidc.go
Comment (tpohl) [reply providers/keycloak_oidc.go:61]: Hi, I understand your concerns and it might pose a risk of unintended behavior.
Still I think that especially for Keycloak there is a difference between group membership and roles (realm and client level).

In our usecase we use roles for permissions -> for this proxy this means „you can pass or you shall not pass“ and we use groups as a manner of managing users (all members of a group have the same permissions but multiple groups can have those permissions).

I have seen other users and implementations that use the „groups“ for permissions.

On top of that at least Keycloak will allow us to design the tokens the way we want to… so there is a lot that can be done wrong.

The implementation of @pb82 leverages the default behavior of Keycloak and where it would map groups and roles into the token. So I think for Keycloak it is alright to use „role:“ as a delimiter. But: Keycloak does allow me to create a group „role:admin“ and I am not aware of any way around this.
One Idea would be to configure if we want to authorize based upon groups or on roles.

One extr thing, that I could imagine is that, if we went for a completely independent OIDC JWT implementation we could configure a JSON Path to the claim in the token, we want to verify and then configure what values would be required to let the user pass.

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: Good call out @JoelSpeed!

While this isn't externally forceable, if a misconfiguration did occur this would be an authz bypass vulnerability for some installations.

I want to figure out a way where the simple behavior in most provider usages that just use `group` don't have a `group:blah` prefix in every group in the `X-Forwarded-Groups` header. Not forcing the `group` canonical use case if a small percentage of the sysadmins use a corner case authz mechanism (e.g. Keycloak roles, or GitLab projects).

Maybe we figure out a scheme that is clear, but uses characters that are very likely not allowed in group names (or wouldn't be common)?

Maybe borrowing type casting syntax? `(role)admin` instead of `role:admin`?

Comment (JoelSpeed) [reply providers/keycloak_oidc.go:61]: I had wondered if we set some sort of option on the OIDC provider to prefix or not to prefix. Then where we have specific providers that have these multiple types, groups, roles, projects, that provider would say it needs groups prefixed and that would affect all installations for that provider. 

To me that's a reasonable limitation when using a provider that supports multiple types of grouping. The usage would just be that when using allowed groups you have to set `<type>:<name>` always on certain providers (we could have validation for this)

I haven't looked at the code though, @NickMeves do you think this would be achievable?

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: > I had wondered if we set some sort of option on the OIDC provider to prefix or not to prefix.

That's a good idea, would potentially help with each sub-provider that has customizations needing to override `Authorization` based logic (I think GitLab needed to do this). Vs sticking with pure no-prefix OIDC group logic, I could keep the code simple and not override default implementations.

Doesn't help the header bloat with people who only use `groups` now having `X-Forwarded-Groups: group:foo,group:bar` -- but maybe that's just cosmetic and isn't too annoying...?

My fear: It might impact OOTB vendor software that takes in comma-delimited group names from a header to internal RBAC mappings/assignments (the original intent of `X-Forwarded-Groups`) with `group:` prefixed 🤔

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: Or crazy idea we ditch the monolithic `session.Groups` and start splitting off various grouping concepts in each provider into their own named header: `X-Forwarded-Groups`, `X-Forwarded-Roles`, `X-Forwarded-Projects`, `X-Forwarded-Repos`, etc...

But that's a much bigger change 😰

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: > My fear: It might impact OOTB vendor software that takes in comma-delimited group names from a header to internal RBAC mappings/assignments (the original intent of `X-Forwarded-Groups`) with `group:` prefixed 🤔

My wonky type casting idea would definitely break this. We should stick with `prefix:name` style to be compatible with downstream consuming syntax.

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: Other option is warning log messages -- if we detect an un-prefixed group that matches a registered alternative for a given provider, we emit a warning log message?

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: @JoelSpeed -- I think I've talked myself into leaving the formatting as is and not punishing 99% of users for 1% who might make weird configuration choices that cause conflicts (which is in their power as admins to remediate). Especially since this is purely an admin setup issue, and can't be forced maliciously from client-side (hence no large security issue).

I think the user experience for people who use backend systems who will take groups from headers and convert to backend systems' RBAC is much better if those backend systems don't have `group:` in every name assuming default `group` structure from OIDC IDTokens is the primary use case.

Especially if the backend system doesn't support `:` in their group/role structure, which is likely a thing in some COTS systems, we are potentially causing a bug that they can't fix.

Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: (This is my long-winded way of saying if I didn't recuse myself, I would approve and merge this PR 😄 )

File: providers/keycloak_oidc_test.go
Comment (NickMeves) [inline providers/keycloak_oidc_test.go:134]: Does this test both Realm Access & Resource Access (two separate helper methods) to get roles?

Since I only see `role:default_read` added, I'm assuming the payload const from above only has 1?

Is there a way to structure that test access token to be clearer as to what we're testing? E.g. building it on the fly via claims in a test table rather than base64 encoding something and setting that as a const which is hard to discern the logic?

File: providers/keycloak_oidc_test.go
Comment (NickMeves) [inline providers/keycloak_oidc_test.go:11]: I'm actually surprised out linter didn't yell at this for the import formatting, it's usually so particular. Can you ditch these extra newlines?

File: CHANGELOG.md
Comment (NickMeves) [inline CHANGELOG.md:7]:
```suggestion
- New keycloak-oidc provider added with support for role based authentication. Will eventually deprecate the current keycloak auth provider.
```
```suggestion
- [#1210](https://github.com/oauth2-proxy/oauth2-proxy/pull/1210) New `keycloak-oidc` provider added with support for role based authentication. Will eventually deprecate the current keycloak auth provider.
```

**Comment by NickMeves** [comment]
I just realized this has a bug while working a similar refactor for the `GitLabProvider`:

During `RefreshSessionIfNeeded`, the new IDToken's `groups` claim will clobber the prexisting `session.Groups` that might've been enriched with Roles from the Access Token (and feasibly the refreshed access token might have updated roles?).

https://github.com/oauth2-proxy/oauth2-proxy/pull/1239#issuecomment-864472811

This might need a refactor on how the formerly private `createSession` worked under the hood in OIDCProvider so IDTokens & Access Tokens are appropriately handled in the redeem & refresh stages (`EnrichSession` is post login only).

**Comment by pb82** [comment]
@NickMeves thanks for the feedback! I'll address the suggested changes and take a look at your commit in #1239 regarding the enrichSession bug.

**Comment by NickMeves** [comment]
> @NickMeves thanks for the feedback! I'll address the suggested changes and take a look at your commit in #1239 regarding the enrichSession bug.

You likely just need to implement `RefreshSessionIfNeeded` wrapping the `OIDCProvider` version call with a final additional call to `extractRoles` to get any updated roles from the new Access Token after the refresh.

**Comment by pb82** [comment]
@NickMeves @JoelSpeed sorry for the delay. I've updated the test payload so that it also includes realm_access roles.

I can't see any use of `refreshSessionIfNeeded` in the gitlab provider, looks this has moved into stored_session.go?

File: providers/keycloak_oidc_test.go
Comment (pb82) [reply providers/keycloak_oidc_test.go:131]: I've added a realm_access role to the payload and updated the test.

**Comment by NickMeves** [comment]
> @NickMeves @JoelSpeed sorry for the delay. I've updated the test payload so that it also includes realm_access roles.
> 
> I can't see any use of `refreshSessionIfNeeded` in the gitlab provider, looks this has moved into stored_session.go?

The interface was refactored. The providers now use `RefreshSession` and always refresh if called. Outer management layers deal with conditional logic to refresh or not.

**Comment by pb82** [comment]
@NickMeves I see. Is there anything else I should address in this PR?

**Review by timgoud — DISMISSED** [review]

**Comment by pb82** [comment]
@timgoud @JoelSpeed I've rebased, this PR. Ok to run the workflows and merge?

File: CHANGELOG.md
Comment (JoelSpeed) [inline CHANGELOG.md:12]: Just want to reword this a bit so it is more actionable
```suggestion
  - [#1210](https://github.com/oauth2-proxy/oauth2-proxy/pull/1210) A new `keycloak-oidc` provider has been added with support for role based authentication. The existing keycloak auth provider will eventually be deprecated and removed. Please switch to the new provider `keycloak-oidc`.
```

File: docs/docs/configuration/auth.md
Comment (JoelSpeed) [inline docs/docs/configuration/auth.md:153]: This will highlight it as a note in the rendered documentation
```suggestion
:::note this is the legacy Keycloak Auth Prodiver, use `keycloak-oidc` if possible. :::
```

File: docs/docs/configuration/overview.md
Comment (JoelSpeed) [inline docs/docs/configuration/overview.md:195]: This should include a note about only working with `keycloak-oidc`

File: pkg/apis/options/legacy_options.go
Comment (JoelSpeed) [inline pkg/apis/options/legacy_options.go:669]: The keycloak provider doesn't make use of this does it?

File: providers/keycloak_oidc.go
Comment (JoelSpeed) [reply providers/keycloak_oidc.go:61]: Ok, I'll let this one stay as is for now, if it ever comes up that this needs to be fixed, you're on the hook for it 😉

**Comment by pb82** [comment]
fixed a test, `golangci-lint run` is now passing locally for me.

**Review by JoelSpeed — APPROVED** [review]
LGTM, @NickMeves do you want to give this a final look or are you happy with my review?

**Comment by pb82** [comment]
@JoelSpeed all feedback addressed, thanks for the review! If we run into trouble with the role prefix, i'm happy to refactor it to a different prefix or identifier.

File: providers/keycloak_oidc.go
Comment (NickMeves) [reply providers/keycloak_oidc.go:61]: I've hardened my position after another thread on the GitLab provider brought up how this `group:` prefix broken htpasswd users in odd ways (they had to hack in the `group:` prefix I think). 😄

**Review by NickMeves — CHANGES_REQUESTED** [review]
> I just realized this has a bug while working a similar refactor for the `GitLabProvider`:
> 
> During `RefreshSessionIfNeeded`, the new IDToken's `groups` claim will clobber the prexisting `session.Groups` that might've been enriched with Roles from the Access Token (and feasibly the refreshed access token might have updated roles?).
> 
> [#1239 (comment)](https://github.com/oauth2-proxy/oauth2-proxy/pull/1239#issuecomment-864472811)
> 
> This might need a refactor on how the formerly private `createSession` worked under the hood in OIDCProvider so IDTokens & Access Tokens are appropriately handled in the redeem & refresh stages (`EnrichSession` is post login only).

Sorry for the delay! I was off the grid for a while 😅 

I'm not seeing this issue addressed: There needs to be an override of RefreshSession that calls the extractRoles logic.  Otherwise anyone that configures session refreshes will have any role details clobbered in the refresh and not re-extracted from the new access token.

**Comment by NickMeves** [comment]
I ...think... its as simple as a `RefreshSession` equivalent of this:

```
// EnrichSession is called after Redeem to allow providers to enrich session fields
// such as User, Email, Groups with provider specific API calls.
func (p *KeycloakOIDCProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
	err := p.OIDCProvider.EnrichSession(ctx, s)
	if err != nil {
		return fmt.Errorf("could not enrich oidc session: %v", err)
	}
	return p.extractRoles(ctx, s)
}
```

**Comment by pb82** [comment]
@NickMeves apologies, I looked for `RefreshSession` in the code but could not find it being used anywhere. I saw now that its used indirectly. Anyway, i've added `RefreshSession` to the keycloak_oidc provider and also a test for it.

**Review by NickMeves — APPROVED** [review]
Thanks for getting this over the finish line! LGTM!

**Pull request merged**
